### ISSUE TITLE
Introduce a wildcard type parameter for visit(NodeList n, A arg) in VoidVisitor and GenericVisitor

### ISSUE THREAD

**Issue by malteskoruppa** [javaparser/javaparser#1271]
I have a class that implements the VoidVisitor interface. Currently, I have something like this in my code:

```
@Override
public void visit(NodeList nodes, Arg a) {
  for (Node node : (NodeList<?>) nodes) {
    // ...
  }
}
```

This is a bit ugly, I would prefer to have this:
```
@Override
public void visit(NodeList<?> nodes, Arg a) {
  for (Node node : nodes) {
    // ...
  }
}
```

But, of course, I have to implement the method specified in the interface. Could the VoidVisitor and the GenericVisitor not prescribe the method `visit(NodeList<?> n, A arg)` (or maybe `visit(NodeList<? extends Node> n, A arg)`... not sure what the best practice is here) instead of `visit(NodeList n, A arg)`?

I have also had a look at a few of the visitors provided by JavaParser itself. For instance, the HashCodeVisitor currently does this:
```
public Integer visit(NodeList n, Void arg) {
  int result = 0;
  for (Object node : n) {
    result += 31 * ((Visitable) node).accept(this, arg);
  }
  return result;
}
```
Wouldn't it be nicer if it could do this? :smile: 
```
public Integer visit(NodeList<?> n, Void arg) {
  int result = 0;
  for (Node node : n) {
    result += 31 * node.accept(this, arg);
  }
  return result;
}
```

This is only a suggestion, I'd be interested to know what you think.

**Comment by matozoid** [comment]
The answer is disappointingly short: check out #1056 :-/ (Although a fix is scheduled for 4.0.0)

**Comment by malteskoruppa** [comment]
Ok. I'm happy to hear a fix is scheduled for 4.0.0. Implementing this change in 3.x would break backwards compatibility with earlier 3.x releases anyway, I suppose you don't want to do that (I know I wouldn't).

Ah, I wish I had the time to experiment with this and perhaps contribute a fix. :grinning: Alas! No land in sight for the next a few months, I have so much work elsewhere. :-/
